xcdat/temporal_avg.py

```diff
@@ -0,0 +1,986 @@
+                >>> ]
+
+        Returns
+        -------
+        xr.Dataset
+            Dataset containing the averaged data variable.
+
+        References
+        ----------
+        .. [1] https://github.com/pydata/xarray/issues/810
+
+        Examples
+        --------
+        Import temporal averaging functionality:
+
+        >>> import xcdat
+
+        Access temporal averaging method:
+
+        >>> # First option
+        >>> ds.temporal.temporal_avg(...)
+        >>> # Second option
+        >>> ds.xcdat.temporal_avg(...)
+
+        Get weighted temporal average for a variable:
+
+        >>> ds_year = ds.temporal.temporal_avg(
+        >>>     "timeseries",
+        >>>     "year",
+        >>>     data_var="ts"
+        >>> )
+        >>> ds_year.ts
+
+        Get weighted temporal average for a variable ("season" frequency):
+
+        >>> ds_season_with_djf = ds.temporal.temporal_avg(
+        >>>     "timeseries",
+        >>>     "season",
+        >>>     data_var="ts",
+        >>>     season_config={
+        >>>         "dec_mode": "DJF",
+        >>>         "drop_incomplete_season": True
+        >>>     }
+        >>> )
+        >>> ds_season_with_djf.ts
+        >>>
+        >>> ds_season_with_jfd = ds.temporal.temporal_avg(
+        >>>     "timeseries",
+        >>>     "season",
+        >>>     data_var="ts",
+        >>>     season_config={"dec_mode": "JFD"}
+        >>> )
+        >>> ds_season_with_jfd.ts
+
+        Get temporal average for a variable ("season" frequency, custom seasons):
+
+        >>> custom_seasons = [
+        >>>     ["Jan", "Feb", "Mar"],  # "JanFebMar"
+        >>>     ["Apr", "May", "Jun"],  # "AprMayJun"
+        >>>     ["Jul", "Aug", "Sep"],  # "JunJulAug"
+        >>>     ["Oct", "Nov", "Dec"],  # "OctNovDec"
+        >>> ]
+        >>>
+        >>> ds_season_custom = ds.temporal.temporal_avg(
+        >>>     "climatology",
+        >>>     "season",
+        >>>     data_var="ts",
+        >>>     season_config={"custom_seasons": custom_seasons}
+        >>> )
+
+        Get data variable's temporal average operation attributes:
+
+        >>> ds_season_with_djf.ts.attrs
+        {
+            'operation': 'temporal_avg',
+            'mode': 'climatology',
+            'freq': 'season',
+            'groupby': 'season',
+            'weighted': 'True',
+            'centered_time': 'True',
+            'dec_mode': 'DJF',
+            'drop_incomplete_djf': 'True'
+        }
+        """
+        self._set_obj_attrs(mode, freq, weighted, centered_time, season_config)
+
+        # Perform Dataset level operations.
+        ds = self._dataset.copy()
+        if self._centered_time:
+            ds = self.center_time(ds)
+
+        # Perform data variable level operations.
+        dv = get_data_var(ds, data_var)
+        ds[dv.name] = self._averager(dv)
+
+        # Keep the original data var for calculating departures.
+        if mode == "climatology":
```

Comment:
I'm not sure what the best solution is here. I think it may be better to leave the variable (e.g., `tas`) as-is and add a new variable (e.g., `tas_climatology`). This seems more intuitive to me, but it would be useful to hear others' thoughts. I'm not far enough in to see what the logic is for departures, but this might help streamline the logic in that function (you'd likely need to add an argument specifying the departure type, monthly or seasonal): 

```
climo_vname = ds.variable_id + '_climatology'
# calculate the climatology if it doesn't exist
if climo_vname not in list(ds.keys()):
    self = self._temporal_avg("climatology", "month")
```

Reply:
Hey Steve, I'll bring this up in the XCDAT meeting to gather input.

> I'm not sure what the best solution is here. I think it may be better to leave the variable (e.g., tas) as-is and add a new variable (e.g., tas_climatology). This seems more intuitive to me, but it would be useful to hear others' thoughts. 

My original thought process from Slack:
- `tas` - the variable that is actively being worked on (climatology, departures, spatial avg, etc.). We must maintain the “original” name ts so that xcdat can properly infer operations using the xcdat_infer attr.
- `tas_original` - the actual original, unmodified variable. Used for calculating departures, which gets stored in ts.

I was thinking if we wanted to support xcdat inference of data variables using the `xcdat_infer` attribute, we need to work on `tas`. One example is if we ran spatial averaging on `tas`, which does not add a new data variable like `tas_spatial_avg`. 
```python
ds = xcdat.open_dataset("tas")

ds.attrs["xcdat_infer"]  # "tas"

# "tas" is inferred from "xcdat_infer"
ds_spatial = ds.xcdat.spatial_avg(...)

# "tas" is inferred from "xcdat_infer"
ds_climo = ds_spatial.xcdat.temporal_avg(...)
```

Despite this, I am now having second thoughts about the `xcdat_infer` system altogether. It seems to be adding feature bloat and an unnecessary overhead when working with the API. It is more straightforward to have the user explicitly pass the key of the data variable, even if it might be redundant for datasets with only one data var. The behavior would at least remain consistent, regardless if the dataset was opened via xarray or XCDAT, or if a single or multiple data vars exist.

```python
ds = xcdat.open_dataset("tas")

# "tas_spatial_avg" gets added, "tas" remains untouched
ds_spatial = ds.xcdat.spatial_avg(data_var="tas")

# "tas_climo" gets added, "tas" and "tas_spatial_avg" remains untouched
ds_climo = ds_spatial.xcdat.temporal_avg(data_var="tas_spatial")
```

> this might help streamline the logic in that function so that you could call departure directly without calculating the climatology first

This answers a thought I had about whether we needed to call climatology first, or if departure should work without calling climatology (I noticed CDAT supports this).